lib/elixir/lib/access.ex

```diff
@@ -31,83 +26,52 @@ defmodule Access do
-  `Access.get(term, key, nil)`.
+  The bracket syntax itself places no restrictions on the key types, but the data
+  structure being accessed may have such restrictions (ie. maps can have any key type,
+  but keyword lists must have atoms as keys).
 
-  `Access` can be combined with `Kernel.put_in/3` to put a value
-  in a given key:
+  This works well for dynamic data structures where keys can come and go over the
+  lifetime of a program _but is disabled by default for structs_, which have keys
+  defined at _compile-time_ (see: [Structs and Dot Syntax](#module-structs-and-dot-syntax)).
 
-      iex> map = %{a: 1, b: 2}
-      iex> put_in map[:a], 3
-      %{a: 3, b: 2}
+  ### Bracket Syntax and Access
 
-  This syntax is very convenient as it can be nested arbitrarily:
+  When compiling  source code, Elixir transforms each invocation of bracket syntax
+  into individual calls to `Access.get/2`.
 
-      iex> users = %{"john" => %{age: 27}, "meg" => %{age: 23}}
-      iex> put_in users["john"][:age], 28
-      %{"john" => %{age: 28}, "meg" => %{age: 23}}
+      iex> quote do a[:b] end
+      {{:., [], [Access, :get]}, [], [{:a, [], AccessTest}, :b]}
 
-  Furthermore, `Access` transparently ignores `nil` values:
+  Because `Access.get/2` is translated into `Access.get/3` with `nil` as the default,
+  accessing a non-existing key will _not raise_, but will instead return a `nil` value.
 
       iex> keywords = [a: 1, b: 2]
       iex> keywords[:c][:unknown]
       nil
 
-  Since `Access` is a behaviour, it can be implemented for key-value
-  data structures. The implementation should be added to the
-  module that defines the struct being accessed. `Access` requires the
-  key comparison to be implemented using the `===` operator.
-
-  ## Static lookups
-
-  The `Access` syntax (`data[key]`) cannot be used to access fields in
-  structs, since structs do not implement the `Access` behaviour by
-  default. It is also a design decision: the dynamic access lookup
-  is meant to be used for dynamic key-value structures, like maps
-  and keywords, and not by static ones like structs (where fields are
-  known and not dynamic).
-
-  Therefore Elixir provides a static lookup for struct fields and for atom
-  fields in maps. Imagine a struct named `User` with a `:name` field.
-  The following would raise:
+  Structs do not by default implement Access behaviour and so by default cannot use this
+  notation (see: [Structs and Dot Syntax](#module-structs-and-dot-syntax)) but
+  Elixir allows the Access behaviour to be [implemented](#module-implementing-the-behaviour)
+  for user-defined struct types after which, bracket syntax may be used.
 
-      user = %User{name: "John"}
-      user[:name]
-      # ** (UndefinedFunctionError) undefined function User.fetch/2 (User does not implement the Access behaviour)
-
-  Structs instead use the `user.name` syntax to access fields:
-
-      user.name
-      #=> "John"
+  ## Access and `Kernel`
```

Comment:
This is challenging. Because we have two syntaxes, bracket and dot (still calling it “dot syntax” for now even though I like the idea of calling it “field” access) and it’d be great to talk about them together, but as far as I can tell, dot-syntax isn’t really related to the Access module so it’s hard to know the right place to put it.

One thought I had is that we could relocate the discussion of the syntaxes into: lib/elixir/pages/Syntax Reference.md (the ‘Access’ section) and keep this focused the core Access module functions, behavior, and helper functions (for Kernel).

For detail about the Kernel accessors (and how they interpret both kinds of syntax), we could put that in `Kernel`.

Here’s the summary of that idea:

- Elixir Syntax Reference
    - Bracket syntax
       - Designed for dynamic data structures
           - Note about compiles to Access.get/2, etc...
           - Link to Access module documentation
    - Dot Syntax
        - Designed for structs with fields
        - Discussion of structs and how it raises
- Access module
    - Detail about the 4 functions underlying bracket syntax and Kernel accessor functions
        - Link to discussion in syntax document
    - Detail and examples about helper functions (Access.all/0) and how they can be used as keys with Kernel accessor functions
        - Link to Kernel accessor functions
- Kernel
    - Accessor functions (`put_in`, etc...)
        - Can be used with anything that implements Access behaviour
            - Link to Access module documentation
        - Detail on how you can use a function as a key
            - Link to Access module documentation for the off-the-shelf helper functions that can be used as keys
        - Includes macros that interpret both dot syntax and bracket syntax
            - Link to syntax in main Elixir syntax discussion

Do you think that’s an improvement?